Monitoring class for process data collection

jseg380 opened the pull request:
# Description

This PR includes the development done to implement `Monitoring class for process data collection`, as well as the documentation related to its use.

Related to https://github.com/wazuh/wazuh/issues/https://github.com/wazuh/wazuh/issues/24683

Commit:
Initial upload

jseg380 committed:
style: remove comment

Commit:
fix: replaced dot characters from reports

Commit:
refactor: delete unnecessary tests

Commit:
style: removed unnecessary quotes and tmp file

Commit:
fix: added specific folder for files CSV

Added Bytes (B) as unit of disk usage tracking. Removed optional type of
parameter in method _get_partition_for_path.

Commit:
change: removed linter references

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md:
````diff
@@ -0,0 +1,157 @@
+
+```shell script
+# Create a virtual environment
+python -m venv virtualenv
````
```suggestion
python3 -m venv virtualenv
```

The command should be `python3` instead of `python` (it also depends on your Python configuration).

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/49d6acc9edd00858050339c832c1025a1b2003bd

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md:
```diff
@@ -0,0 +1,157 @@
+source virtualenv/bin/activate
+
+# Install the package using pip
+python -m pip install .
```
```suggestion
python3 -m venv virtualenv
```

Same as above. Change `python` to `python3`.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/49d6acc9edd00858050339c832c1025a1b2003bd

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md:
```diff
@@ -0,0 +1,157 @@
+python -m pip install .
+
+# Verify the correct installation
+python -m pip list | grep process_resource_monitoring
```
Same as above. Change `python` to `python3`.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/49d6acc9edd00858050339c832c1025a1b2003bd

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/__init__.py:
```diff
@@ -0,0 +1,16 @@
+monitor -- Process and child processes monitoring using one process per
+           instance (one thread). Tracks several resource usage metrics
+
+disk_usage_tracker -- Track disk usage of files/directories over time. Show 
```
```suggestion
disk_usage_tracker -- Track disk usage of files/directories over time. Show
```

Remove trailing whitespace.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/__init__.py:
```diff
@@ -0,0 +1,16 @@
+"""Process resource usage monitoring tool.
+
+This package contains the following modules:
```
```suggestion
# Copyright (C) 2015-2021, Wazuh Inc.
# Created by Wazuh, Inc. <info@wazuh.com>.
# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2

"""Process resource usage monitoring tool.

This package contains the following modules:
```

Add copyright.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/__init__.py:
```diff
@@ -0,0 +1,16 @@
+                      absolute and relative (to the partition) usage.
+
+"""
+
```
```suggestion


```

Add blank space

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/__init__.py:
```diff
@@ -0,0 +1,16 @@
+
+from process_resource_monitoring.disk_usage_tracker import DiskUsageTracker
+from process_resource_monitoring.monitor import Monitor
+
```
```suggestion


```

Add blank space.

jseg380 replied:
Not sure if the convention we are following instructs to leave 2 blank lines after the import blocks but with the current configuration I am using for the linter, it is marked as an error.
```
src/process_resource_monitoring/__init__.py:18:1: I001 [*] Import block is un-sorted or un-formatted
   |
18 | / from process_resource_monitoring.disk_usage_tracker import DiskUsageTracker
19 | | from process_resource_monitoring.monitor import Monitor
20 | | 
21 | | 
22 | | __all__ = ['Monitor', 'DiskUsageTracker']
   | |_^ I001
   |
   = help: Organize imports

Found 1 error.
[*] 1 fixable with the `--fix` option.
```

rafabailon replied:
We have been using that style in other PRs. Even so, since the styles are not completely defined, I think we can leave it like that and if @Rebits  thinks that the change should be made, he will let you know in his review.

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/_logger.py:
```diff
@@ -0,0 +1,7 @@
+"""Logger instance shared by all the modules to track workflow."""
```
```suggestion
# Copyright (C) 2015-2021, Wazuh Inc.
# Created by Wazuh, Inc. <info@wazuh.com>.
# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2

"""Logger instance shared by all the modules to track workflow."""
```

Add copyright.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/_logger.py:
```diff
@@ -0,0 +1,7 @@
+"""Logger instance shared by all the modules to track workflow."""
+
```
```suggestion


```

Add blank space.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/_logger.py:
```diff
@@ -0,0 +1,7 @@
+"""Logger instance shared by all the modules to track workflow."""
+
+import logging
+
```
```suggestion


```

Add blank space

jseg380 replied:
Same as [here](https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1768913070)

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,228 @@
+usage of one or multiple files/directories, showing the space it takes
+relative to the partition in which it is located.
+"""
+
```
```suggestion


```

Add blank space

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,228 @@
+            path (str): Path of the file.
+
+        Returns:
+            (int): Size of the file in the unit specified in the constructor.
```
In the `Returns` it is specified that the returned value is an `int` but in the method definition it appears as `float`. They should be the same.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/commit/888fbfbd964f340da6ea47fe116073b3fbf87a97

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,228 @@
+        return self.get_file_size(path) / partition_size
+
+    def is_event_set(self) -> bool:
+        """Check if the internal flag for the tracker event is set."""
```
The return is missing in the documentation.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/commit/888fbfbd964f340da6ea47fe116073b3fbf87a97

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/monitor.py:
```diff
@@ -0,0 +1,284 @@
+ - Disk_Read
+ - Disk_Written
+"""
+
```
```suggestion


```

Add blank space

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/monitor.py:
```diff
@@ -0,0 +1,284 @@
+        return info
+
+    def is_event_set(self) -> bool:
+        """Check if the internal flag for the Monitor event is set."""
```
The return is missing in the documentation.

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/888fbfbd964f340da6ea47fe116073b3fbf87a97

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/wazuh_metrics.py:
```diff
@@ -0,0 +1,333 @@
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
+"""Script to monitor Wazuh central components resource usage."""
+
```
Add blank space

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/42bdbf6a440e7627707ed2e0df6b151f843ec8f1

rafabailon commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/wazuh_metrics.py:
```diff
@@ -0,0 +1,333 @@
+from process_resource_monitoring._logger import logger
+from process_resource_monitoring.disk_usage_tracker import DiskUsageTracker
+from process_resource_monitoring.monitor import Monitor
+
```
Add blank space

jseg380 replied:
Same as [here](https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1768913070). 
It seems like if there is a class after the import block two spaces are required, otherwise only one.

rafabailon reviewed (changes requested):
Great Job! I've tested the code and it seems to work correctly. I've left a few comments, mostly about styles.

Commit:
fix: changed python to python3 in README

Commit:
style: added missing blank lines

Commit:
fix: fixed returns values in docstrings

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+
+
+class DiskUsageTracker:
+    """Class to track disk usage of the Wazuh installation components over time.
```
```suggestion
    """Class to track disk usage of a specified partition over time.
```

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/8085b993690a9b14c1b1ab81b55e29ec02d82270

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+            Start the monitoring threads.
+        shutdown:
+            Stop all the monitoring threads.
+    """
```
Missing methods. Remove this description or restrict public methods to these

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/7349859c18c552b13eeecc88951a495e06389aad

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+
+    _DATA_UNITS: Dict[str, int] = {'B': 0, 'KB': 1, 'MB': 2, 'GB': 3, 'TB': 4}
+
+    def __init__(self, file_path: str, value_unit: str = 'GB', time_step: int = 1, dst_dir: str = gettempdir()) -> None:
```
`dst_dir` should be mandatory. We need to know exactly where is the collected data

jseg380 replied:
It can be known exactly where the collected data is, by default, for both monitor and disk_usage_resource module, if no dst_dir is supplied then it will use this path `tempdir/process_metrics/DD-MM-YYYY/ssssssssss`, where tempdir is the temporal directory (`/tmp` in case of Unix-like systems), `DD-MM-YYYY` is the date the monitoring started, and `ssssssssss` is the moment in time (in seconds since the epoch) when the monitoring started).

Apart from that, in line with https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1773580308, since this script only seems to have been used for tracking process resource usage and the disk monitoring usage was added as a complement, an option could be to have the same path for both (as it's currently made in the `wazuh_metrics.py` script) scripts to save the data to.
Unlike, https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1773569369, this one is quite less of a hussle and seems plausible completing it in a short amount of time.
Also, I believe it's worth mentioning that this aspect was inherited from the original Monitor class, which seems to have been working fine (in that specific aspect) up until now.

Rebits replied:
Discussed in a meeting with @jseg380. Accepted

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+            file_path (str): path of the file/dir.
+            value_unit (str, optional): unit to store the bytes values. Defaults to GB.
+            time_step (int, optional): time in seconds between each scan. Defaults to 1.
+            dst_dir (str, optional): directory to store the CSVs. Defaults to /tmp directory.
```
Include dst_dir validation

jseg380 replied:
See https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1773624670

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+            Returns:
+                (str): formatted time.
+            """
+            return datetime.fromtimestamp(t).strftime('%d/%m/%Y-%H:%M:%S.%f')
```
```suggestion
        def convert_time(time: float) -> str:
            """Convert time in milliseconds since the epoch to %d/%m/%Y-%H:%M:%S.%f.

            Args:
                time (float): time in milliseconds since the epoch.

            Returns:
                (str): formatted time.
            """
            return datetime.fromtimestamp(t).strftime('%d/%m/%Y-%H:%M:%S.%f')
```

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/8085b993690a9b14c1b1ab81b55e29ec02d82270

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+            'Acc_time': datetime.min,
+            'Creat_time': datetime.min
+        }
+
```
0 value is a valid size. For example, if an empty file does not change over time, the registered values should be 0. I suggest changing this to a negative value

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/a189666f273fec45d0d7854d1e5eea9e33fa92c1

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+            info['Usage(%)'] = self.get_disk_usage(self._file_path, self._partition)
+            info['Mod_time'] = convert_time(os.path.getmtime(self._file_path))
+            info['Acc_time'] = convert_time(os.path.getatime(self._file_path))
+            info['Creat_time'] = convert_time(os.path.getctime(self._file_path))
```
```suggestion
            info['Modification_time'] = convert_time(os.path.getmtime(self._file_path))
            info['Access_time'] = convert_time(os.path.getatime(self._file_path))
            info['Creation_time'] = convert_time(os.path.getctime(self._file_path))
```

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/62765614ee3b0c8f16a1bac2395ff9e94ecd6820

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/disk_usage_tracker.py:
```diff
@@ -0,0 +1,233 @@
+            if self._event is not None:
+                self.shutdown()
+        finally:
+            info.update({key: round(value, 2) for key, value in info.items() if isinstance(value, (int, float))})
```
```suggestion
            info.update({key: round(value, 3) for key, value in info.items() if isinstance(value, (int, float))})
```

jseg380 replied:
Changed in https://github.com/wazuh/wazuh-qa/pull/5747/commits/8085b993690a9b14c1b1ab81b55e29ec02d82270

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/monitor.py:
```diff
@@ -0,0 +1,289 @@
+            f'Disk_Written({self._value_unit})': 0.0,
+            f'Disk_Read_Speed({self._value_unit}/s)': 0.0,
+            f'Disk_Write_Speed({self._value_unit}/s)': 0.0,
+        }
```
0 is a valid value. We should consider initializing this dictionary with invalid values to ensure metrics consistency

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/a189666f273fec45d0d7854d1e5eea9e33fa92c1

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/monitor.py:
```diff
@@ -0,0 +1,289 @@
+                info[f'SWAP({self._value_unit})'] = unit_conversion(memory_data.swap)
+                info['FD'] = proc.num_fds()
+
+                if self._platform == 'linux' or platform == 'win32':
```
We need to give full support to macOS also. Even if some fields are OS-specific, we need to give the much value as possible for each OS.

jseg380 replied:
Psutil doesn't seem to support macOS that much. In the docs you can see how the availability for the method that is used in the script's conditional, Process.io_counters() is only available for Linux, BSD, Windows and AIX. Should we go through the trouble to make it work? It would probably take some time, but taking into account that the central components can only be installed on Linux then it seems reasonable to leave out those 4 out of 11 metrics (7 relevant metrics are allegedly still collected).

Rebits replied:
True https://psutil.readthedocs.io/en/latest/#psutil.Process.io_counters, macOS is not supported

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/process_resource_monitoring/monitor.py:
```diff
@@ -0,0 +1,289 @@
+                        write_speed = (info[f'Disk_Written({self._value_unit})'] - self._previous_write) \
+                            / self._time_step
+                        info[f'Disk_Read_Speed({self._value_unit}/s)'] = read_speed
+                        info[f'Disk_Write_Speed({self._value_unit}/s)'] = write_speed
```
These values are correct? They are usefull?

jseg380 replied:
This information doesn't seem useful. Forgot to remove it. Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/4cd84689130f290d5d1cae1c13b0f7aa56a780ec

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/wazuh_metrics.py:
```diff
@@ -0,0 +1,334 @@
+
+"""Script to monitor Wazuh central components resource usage."""
+
+
```
We should consider split this script: one for binary metrics and another for disk space tracker

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/24dc63523c233f94cc7823e263db2baf2ac8c0a3  , https://github.com/wazuh/wazuh-qa/pull/5747/commits/2b839d3564e79bf7dde438a7de96408e8aae106f

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md:
```diff
@@ -0,0 +1,157 @@
+
+## Prerequisites
+
+- Wazuh component(s) 4.9.0 (or greater)
```
Why?

jseg380 replied:
Because, as mentioned in https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1771830749, at the moment of writing this this module is dependant of Wazuh, since it is only able to monitor processes with specific names that Wazuh uses for their processes. See https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1773569369 for the reason behind keeping this a Wazuh-dependent script.

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md:
```diff
@@ -0,0 +1,157 @@
+
+#### Wazuh manager
+
+| Process | Argument |
+| ------- | -------- |
+| wazuh-agentlessd | agentlessd |
+| wazuh-analysisd | analysisd |
+| wazuh_apid.py | apid |
+| wazuh-authd | authd |
+| wazuh_clusterd.py | clusterd |
+| wazuh-csyslogd | csyslogd |
+| wazuh-db | db |
+| wazuh-dbd | dbd |
+| wazuh-execd | execd |
+| wazuh-integratord | integratord |
+| wazuh-logcollector | logcollector |
+| wazuh-maild | maild |
+| wazuh-modulesd | modulesd |
+| wazuh-monitord | monitord |
+| wazuh-remoted | remoted |
+| wazuh-syscheckd | syscheckd |
```
All these processes will change in 5.0.0. We need to make tool independently of the target process we are going to monitor

jseg380 replied:
Some concerns arise from here:
- How would this be achieved? Since, as of right now and following the rationale of "making the tool Wazuh-independent" it should be able to monitor any "common" process, such as a python script (which is actually executed by the Python interpreter). In some cases like the aforementioned one the process can not be found in the name and has to be searched in the process arguments (in Unix-like systems you can use: `ps -e -o args` to see the arguments of the processes). This could prove troublesome since Wazuh processes names make use of extremely common words, leading to potential incorrect process tracking and waste of resources. For example, when looking for processes with the word "db" on it in my machine I get 19 results (1 is the grep itself):
```shell
$ ps -e -O args | grep db | wc -l
20
```
Some examples of processes containing the word "db":
```
PID  COMMAND                        S ?        TIME       ARGS
---  ------------------------------ - -------- ---------- ------------------------------------
###  /usr/lib/systemd/systemd-us     S ?        00:00:00   /usr/lib/systemd/systemd-userdbd
###  /usr/bin/dbus-broker-launch     S ?        00:00:00   /usr/bin/dbus-broker-launch --scope system --audit
###  dbus-broker --log # --contr     S ?        00:00:00   dbus-broker --log # --controller # --machine-id XXXXXXX --max-bytes ######## --max-fds ###### --max-matches ######## --audit
###  /usr/bin/dbus-broker-launch     S ?        00:00:00   /usr/bin/dbus-broker-launch --scope user
###  dbus-broker --log # --contr     S ?        00:00:00   dbus-broker --log # --controller # --machine-id XXXXXXX --max-bytes ######## --max-fds ###### --max-matches ########
###  /usr/bin/gmenudbusmenuproxy     S ?        00:00:02   /usr/bin/gmenudbusmenuproxy
```

- Is it worth the effort and time? In my opinion, no. Here are some points to support my opinion:
  - It seems like this script was made for a very specific task and currently there doesn't seem to be any prospect of being used in other scenarios.
  - Making a real "generic/universal" process resource usage tracking program would take a considerable amount of time since there are many scenarios to be taken into account, and let's not forget that it would have to support several OS (Linux, Windows and MacOS in the very least, since all of these ones can have and agent installed, but so do Solaris, AIX, HP). Even though Python aims to be an OS-independent programming language, it can't avoid it (see the conditional in the `monitor.py` module).
  - Due to the ever-changing nature of software development it is virtually impossible to make a product that is completely future-proof. Pursuing this goal may result in a huge waste of time and resources, since it would have to be modified in the future anyway, either for maintenance or improvements.
  - After saying all this it seems feasible to me that modifying the current script and/or making a new script  from scratch whenever it is needed, could be less time-consuming than the alternative.
  - Other problems that arise from making a full-blown program is that new concerns would carry new concerns to take into account, such as performance, stability, etc.

Overall, it does not seem worth it (at least for now). This is just my honest opinion, suggestions are welcome.

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md:
```diff
@@ -0,0 +1,157 @@
+- **Usage:** Percentage of the space the file takes relative to the partition's size.
+- **Mod_time:** Last time the file was modified.
+- **Acc_time:** Last time the file was accessed.
+- **Creat_time:** Creation time (Windows) or metadata change time (Unix).
```
Don't use abbreviations

jseg380 replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5747/commits/62765614ee3b0c8f16a1bac2395ff9e94ecd6820

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/wazuh_metrics.py:
```diff
@@ -0,0 +1,334 @@
+# Constant variables
+METRICS_FOLDER: str = join(gettempdir(), 'process_metrics')
+CURRENT_SESSION: str = join(METRICS_FOLDER, datetime.now().strftime('%d-%m-%Y'), str(time_ns() // 10**9))
+PROCESS_MAPPING: Dict[str, str] = {
+    # Manager processes
+    'agentlessd': 'wazuh-agentlessd',
+    'analysisd': 'wazuh-analysisd',
+    'apid': 'wazuh_apid.py',
+    'authd': 'wazuh-authd',
+    'clusterd': 'wazuh_clusterd.py',
+    'csyslogd': 'wazuh-csyslogd',
+    'db': 'wazuh-db',
+    'dbd': 'wazuh-dbd',
+    'execd': 'wazuh-execd',
+    'integratord': 'wazuh-integratord',
+    'logcollector': 'wazuh-logcollector',
+    'maild': 'wazuh-maild',
+    'modulesd': 'wazuh-modulesd',
+    'monitord': 'wazuh-monitord',
+    'remoted': 'wazuh-remoted',
+    'wazuh-syscheckd': 'syscheckd',
+
+    # Indexer
+    'indexer': 'wazuh-indexer',
+
+    # Dashboard
+    'dashboard': 'wazuh-dashboard',
+
+    # Agent
+    'agentd': 'wazuh-agentd',
+}
```
Why enforce the processes?

jseg380 replied:
Following the rationale of https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1773569369 chances are it will be "better" (less time, resources) to simply take a "naive approach" and modify it whenever it is needed. To address differences between versions, each branch/tag would have the script with the correct parameters. There are other ways of handling so, for example, having a mapping with all the relevant processes across versions and enforce the one specific to only one version.

Rebits replied:
Resolved in a meeting with @jseg380

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/src/wazuh_metrics.py:
```diff
@@ -0,0 +1,334 @@
+    healthy = True
+    for process, monitors in active_monitors.items():
+        # Check if there is any unhealthy monitor
+        if any(filter(lambda monitor: monitor.is_event_set(), monitors)):
+            logger.warning(f'Monitoring of {process} failed. Attempting to create new monitor instances')
+
+            try:
+                # Try to get new PIDs
+                process_pids = Monitor.get_process_pids(process)
+                # Shutdown all the related monitors to the failed process (necessary for multiprocessing)
+                for monitor in monitors:
+                    monitor.shutdown()
+            except ValueError:
+                healthy = False
+                logger.warning(f'Could not create new monitor instances for {process}')
+                continue
+
+            for i, pid in enumerate(process_pids):
+                # Attempt to create new monitor instances for the process
+                p_name = process if i == 0 else f'{process}_child_{i}'
+                monitor = Monitor(
+                    process_name=p_name,
+                    pid=pid,
+                    value_unit=options.data_unit,
+                    time_step=options.sleep_time,
+                    version=options.version,
+                    dst_dir=options.store_process_path,
+                )
+                monitor.start()
+
+                try:
+                    # Replace old monitors for new ones
+                    active_monitors[process][i] = monitor
+                except IndexError:
+                    active_monitors[process].append(monitor)
```
In the case of the dynamic amount of threads used, we will miss extra daemons launched in the middle of the scan. We need to handle that case

jseg380 replied:
Working on it, currently added https://github.com/wazuh/wazuh-qa/pull/5747/commits/067d45b0e0aae6724dbc28d326d44d908cba4494

jseg380 replied:
Finished it on https://github.com/wazuh/wazuh-qa/pull/5747/commits/c9572944b9012dbf636f2c00ce82012abb06edac

Rebits reviewed (changes requested):
Great Job! 

Some minor changes are requested.

***

Remember that some unitary tests are mandatory for new development

Rebits commented on deps/wazuh_testing/wazuh_testing/process_resource_monitoring/README.md:
```diff
@@ -0,0 +1,157 @@
+| wazuh-remoted | remoted |
+| wazuh-syscheckd | syscheckd |
+
+> Note:  
+> `wazuh_apid.py` and `wazuh_clusterd.py` are scripts run by the Python interpreter, not processes themselves.
+
```
This will change in the future. We need to make the package not dependant of wazuh component for maintenance

jseg380 replied:
See https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1773569369 and https://github.com/wazuh/wazuh-qa/pull/5747#discussion_r1773580308

Rebits replied:
Resolved in meeting

Commit:
docs: added missing public methods

Commit:
style: removed explicit Wazuh references

Removed specific references of Wazuh installation to make it more generic. Changed format of numbers in CSV to have 3 decimal values. Modified the percentage of the disk to be in [0.0, 100.0] instead of [0.0, 1.0]

Commit:
docs: do not use abbreviations

Commit:
refactor: removed unused variables

Commit:
fix: set invalid values to detect failure

Commit:
feat: splitted script into separate scripts

Commit:
docs: splitted script into separate scripts

Commit:
docs: fixed invalid md reference

Commit:
change: remove unnecessary instance attribute

Commit:
feat: track processes that spawn post-startup

Commit:
fix: uncaught ValueError caused early finish

Commit:
docs: fixed docstrings and type hinting

Rebits commented:
During the final review, several issues were detected regarding the current approach

## Not handling of exception

### Subprocess is killed during the monitoring
During the monitoring of a multiprocess program, it appears that failures
```trace
Traceback (most recent call last):
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1717, in wrapper
    return fun(self, *args, **kwargs)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_common.py", line 508, in wrapper
    raise raise_from(err, None)
  File "<string>", line 3, in raise_from
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_common.py", line 506, in wrapper
    return fun(self)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1780, in _parse_stat_file
    data = bcat("%s/%s/stat" % (self._procfs_path, self.pid))
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_common.py", line 851, in bcat
    return cat(fname, fallback=fallback, _open=open_binary)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_common.py", line 839, in cat
    with _open(fname) as f:
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_common.py", line 799, in open_binary
    return open(fname, "rb", buffering=FILE_READ_BUFFER_SIZE)
FileNotFoundError: [Errno 2] No such file or directory: '/proc/42457/stat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/__init__.py", line 355, in _init
    self.create_time()
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/__init__.py", line 757, in create_time
    self._create_time = self._proc.create_time()
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1717, in wrapper
    return fun(self, *args, **kwargs)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1948, in create_time
    ctime = float(self._parse_stat_file()['create_time'])
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1726, in wrapper
    raise NoSuchProcess(self.pid, self._name)
psutil.NoSuchProcess: process no longer exists (pid=42457)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/process_resource_monitoring/monitor.py", line 244, in set_process
    self._proc = psutil.Process(self._pid)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/__init__.py", line 319, in __init__
    self._init(pid)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/__init__.py", line 368, in _init
    raise NoSuchProcess(pid, msg=msg)
psutil.NoSuchProcess: process PID not found (pid=42457)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/rebits/Wazuh/monitor/bin/wazuh-process-metrics", line 8, in <module>
    sys.exit(main())
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/wazuh_process_metrics.py", line 309, in main
    monitors_healthcheck(options, process_list)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/wazuh_process_metrics.py", line 221, in monitors_healthcheck
    if check_monitors_health(options):
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/wazuh_process_metrics.py", line 193, in check_monitors_health
    monitor = Monitor(
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/process_resource_monitoring/monitor.py", line 100, in __init__
    self.set_process()
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/process_resource_monitoring/monitor.py", line 246, in set_process
    raise ValueError(f'The process {self._process_name} is not running.') from err
ValueError: The process wazuh_apid.py_child_1 is not running.
```

The tool should be capable of handling this

### The presence of zombie process makes the tool fail

If a zombie process is present in the system, the whole monitoring fails:
```trace
Traceback (most recent call last):
  File "/home/rebits/Wazuh/monitor/bin/wazuh-process-metrics", line 8, in <module>
    sys.exit(main())
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/wazuh_process_metrics.py", line 288, in main
    process_pids = Monitor.get_process_pids(process)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/process_resource_monitoring/monitor.py", line 118, in get_process_pids
    if any(filter(lambda x: f'{process_name}' in x, proc.cmdline())):
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/__init__.py", line 724, in cmdline
    return self._proc.cmdline()
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1717, in wrapper
    return fun(self, *args, **kwargs)
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1856, in cmdline
    self._raise_if_zombie()
  File "/home/rebits/Wazuh/monitor/lib/python3.10/site-packages/psutil/_pslinux.py", line 1761, in _raise_if_zombie
    raise ZombieProcess(self.pid, self._name, self._ppid)
```

## Logging

Tool logging does not differ, contrary to the CSV data, from different iterations of the tool. This should be included in the same directory as the CSV data

## Reports
- Obtained reports are included in a difficult-to-read time-date format
- Several CSVs are created using the same process. Even the child process should be included in the same CSV, otherwise, it would be really hard to handle this data

```shell
➜  Desktop tree -l  /tmp/process_metrics
/tmp/process_metrics
├── 27-09-2024
│   ├── 1727423503
│   │   ├── wazuh_apid_29459.csv
│   │   ├── wazuh_apid_39133.csv
│   │   ├── wazuh_apid_39134.csv
│   │   ├── wazuh_apid_39135.csv
│   │   ├── wazuh_apid_39136.csv
│   │   ├── wazuh_apid_39137.csv
│   │   ├── wazuh_apid_39242.csv
│   │   ├── wazuh_apid_39243.csv
│   │   ├── wazuh_apid_child_1.csv
│   │   ├── wazuh_apid_child_2.csv
│   │   ├── wazuh_apid_child_3.csv
│   │   ├── wazuh_apid_child_4.csv
│   │   ├── wazuh_apid_child_5.csv
│   │   └── wazuh_apid.csv
│   └── 1727425589
│       ├── wazuh_apid_29459.csv
│       ├── wazuh_apid_42457.csv
│       ├── wazuh_apid_42458.csv
│       ├── wazuh_apid_42459.csv
│       ├── wazuh_apid_42460.csv
│       ├── wazuh_apid_42461.csv
│       ├── wazuh_apid_42488.csv
│       ├── wazuh_apid_42489.csv
│       ├── wazuh_apid_child_1.csv
│       ├── wazuh_apid_child_2.csv
│       ├── wazuh_apid_child_3.csv
│       ├── wazuh_apid_child_4.csv
│       ├── wazuh_apid_child_5.csv
│       └── wazuh_apid.csv
└── wazuh-resource-metrics.log
```
